pkg/kubelet/kubelet.go

```diff
@@ -929,34 +929,107 @@ type Kubelet struct {
+	// of desired pods. Note that not all desired pods are running, and not all running pods are in
+	// the podManager - for instance, force deleting a pod from the apiserver will remove it from
+	// the podManager, but the pod may still be terminating and tracked by the podWorkers. Components
+	// that need to know the actual consumed resources of the node or are driven by podWorkers and
+	// the sync*Pod methods	(status, volume, stats) should also consult the podWorkers when
+	// reconciling.
+	//
+	// TODO: review all kubelet components that need the actual set of pods (vs the desired set)
+	// and update them to use podWorkers instead of podManager. This may introduce latency in some
+	// methods, but avoids race conditions and correctly accounts for terminating pods that have
+	// been force deleted or static pods that have been updated.
+	podManager kubepod.Manager
+
+	// podWorkers is responsible for driving the lifecycle state machine of each pod. The worker is
+	// notified of config changes, updates, periodic reconciliation, container runtime updates, and
+	// evictions of all desired pods and will invoke reconciliation methods per pod in separate
+	// goroutines. The podWorkers are authoritative in the kubelet for what pods are actually being
+	// run and their current state:
+	//
+	// * syncing: pod should be running (syncPod)
+	// * terminating: pod should be stopped (syncTerminatingPod)
+	// * terminated: pod should have all resources cleaned up (syncTerminatedPod)
+	//
+	// and invoke the handler methods that correspond to each state. Components within the
+	// kubelet that need to know the phase of the pod in order to correctly set up or tear down
+	// resources must consult the podWorkers.
+	//
+	// Once a pod has been accepted by the pod workers, no other pod with that same UID (and
+	// fullName, for static pods) will be started until the first pod has fully terminated and
+	// been cleaned up by SyncKnownPods. This means a pod may be desired (in API), admitted (in
+	// pod manager), and requested (by invoking UpdatePod) but not start for an arbitrarily long
+	// interval because a prior pod is still terminating.
+	//
+	// As an event-driven (by UpdatePod) controller, the podWorkers must periodically be resynced
+	// by the kubelet invoking SyncKnownPods with the desired state (admitted pods in podManager).
+	// Since the podManager may be unaware of some running pods due to force deletion, the
+	// podWorkers are responsible for triggering a sync of pods that are no longer desired but
+	// must still run to completion.
 	podWorkers PodWorkers
 
+	// evictionManager observes the state of the node for situations that could impact node stability
+	// and evicts pods (sets to phase Failed with reason Evicted) to reduce resource pressure. The
+	// eviction manager acts on the actual state of the node and considers the podWorker to be
+	// authoritative.
```

Comment:
I think eviction manager actually relies on pod manager today, not pod worker. 

1. Eviction manager is started here - https://github.com/kubernetes/kubernetes/blob/703361f381d4a1eaa7dc87ff3b7c12c84f3878a4/pkg/kubelet/kubelet.go and is passed in `kl.GetActivePods`
2. `GetActionPods()` returns pods from pod manager - https://github.com/kubernetes/kubernetes/blob/703361f381d4a1eaa7dc87ff3b7c12c84f3878a4/pkg/kubelet/kubelet_pods.go#L103-L107
3. eviction manager gets pods to evict using the `GetActivePods` - https://github.com/kubernetes/kubernetes/blob/703361f381d4a1eaa7dc87ff3b7c12c84f3878a4/pkg/kubelet/eviction/eviction_manager.go#L257

so at least today, AFAIU the list of pods that eviction manager looks at will be limited to those returned by pod manager, not pod worker.

Reply:
Yeah, I was writing these comments to define who is authoritative (i.e., it doesn't matter how it's implemented today, this is who should be correct).  Before we merge anything I'll address this (either into a TODO, or by fixing it)